README.md
# ONSRaidTools
Introducing the must-have add-on for any serious WoW raider!

 

Developed specifically for the elite guild, One Night Stand on Ravencrest EU, this add-on is the perfect tool for viewing boss positions during raids.

Say goodbye to the hassle of constantly checking raid spreadsheets - with this add-on, all the information you need is right at your fingertips! And best of all, it's incredibly user-friendly.

Simply resize, adjust the opacity, or move the boss positions around to suit your needs.

 

Despite its powerful features, this add-on is impressively lightweight, so you can be sure it won't slow you down during those crucial boss fights. So whether you're a seasoned raider or just starting out, this add-on is an essential tool for your arsenal. Download it today and experience the ultimate in raiding efficiency!


Changelog.md
- First beta release


ONSImageView.lua
local AddOnName, _ = ...
local ONSImageView = LibStub("AceAddon-3.0"):NewAddon(AddOnName, "AceEvent-3.0")


local LDBIcon = LibStub("LibDBIcon-1.0")
local defaults = {
    global = {
        position = {
            x = 0,
            y = 0,
            point = "CENTER",
            relativePoint = "CENTER"
        },
        options = {
            minimap = {
                hide = false
            }
        }
    }
}

local ONSImageViewBroker = LibStub("LibDataBroker-1.1"):NewDataObject(AddOnName, {
    type = "data source",
    icon = "Interface\\Addons\\ONSImageView\\assets\\ons_icon",
    OnClick = function(self, button)
        ONSImageView:ToggleFrame()
    end,
    OnTooltipShow = function(tooltip)
        tooltip:AddLine(AddOnName)
        tooltip:AddLine("Click to toggle the main window")
    end
})


function ONSImageView:OnEnable()
    C_Timer.After(1, function()
        self:ToggleFrame()
    end)
    ONSImageView:RegisterEvent("MODIFIER_STATE_CHANGED")
end

function ONSImageView:OnInitialize()
    self.db = LibStub("AceDB-3.0"):New("ONSImageViewDB", defaults)

    LDBIcon:Register(AddOnName, ONSImageViewBroker, self.db.global.options.minimap)
end

function ONSImageView:UpdateMinimapButton()
    if (self.db.global.options.minimap.hide) then
        LDBIcon:Hide(AddOnName)
    else
        LDBIcon:Show(AddOnName)
    end
end

function ONSImageView:ToggleMinimapButton()
    if (self.db.global.options.minimap.hide) then
        self.db.global.options.minimap.hide = false
    else
        self.db.global.options.minimap.hide = true
    end
    self:UpdateMinimapButton()
end

SLASH_ONS1 = "/ons"

SlashCmdList["ONS"] = function(msg)
    if msg == "settings" or msg == "config" then
        -- TODO: Add options
    elseif msg == "minimap" then
        ONSImageView:ToggleMinimapButton()
    else
        ONSImageView:ToggleFrame()
    end
end


ONSRaidTools.lua
local AddOnName, components = ...
local ONSRaidTools = LibStub("AceAddon-3.0"):NewAddon(AddOnName, "AceEvent-3.0")

local AceConfig = LibStub("AceConfig-3.0")
local AceConfigDialog = LibStub("AceConfigDialog-3.0")


-- default active raid
ONSRaidTools.activeRaid = "DFS2"

local LDBIcon = LibStub("LibDBIcon-1.0")
local defaults = {
    global = {
        position = {
            x = 0,
            y = 0,
            point = "CENTER",
            relativePoint = "CENTER"
        },
        options = {
            minimap = {
                hide = false
            },
            opacity = 1,
            scale = 1
        },
        loadedEncounter = {
            images = {},
            info = {}
        }
    }
}



local ONSRaidToolsBroker = LibStub("LibDataBroker-1.1"):NewDataObject(AddOnName, {
    type = "data source",
    icon = "Interface\\Addons\\ONSRaidTools\\assets\\onsIcon",
    OnClick = function(self, button)
        ONSRaidTools:ToggleFrame()
    end,
    OnTooltipShow = function(tooltip)
        tooltip:AddLine(AddOnName)
        tooltip:AddLine("Click to toggle the main window")
    end
})





local ONSOptionsTable = {
    name = AddOnName,
    type = "group",
    args = {
        description = {
            name = "ONSRaidTools is a tool to help you with your raiding experience.",
            width = "full",
            type = "description",
            order = 0
        },
        -- toggle minimap button
        minimap = {
            name = "Minimap Button",
            width = "full",
            type = "toggle",
            get = function()
                return not ONSRaidTools.db.global.options.minimap.hide
            end,
            set = function(_, value)
                ONSRaidTools.db.global.options.minimap.hide = not value
                ONSRaidTools:UpdateMinimapButton()
            end
        },
        -- opacity slider -> range
        opacity = {
            name = "Opacity",
            width = "full",
            type = "range",
            min = 0.3,
            max = 1,
            step = 0.1,
            get = function()
                return ONSRaidTools.db.global.options.opacity
            end,
            set = function(_, value)
                ONSRaidTools.db.global.options.opacity = value
                ONSRaidTools.mainWindow:SetAlpha(value)
            end
        },
        -- scale slider -> range
        scale = {
            name = "Scale",
            width = "full",
            type = "range",
            min = 0.5,
            max = 2,
            step = 0.1,
            get = function()
                return ONSRaidTools.db.global.options.scale
            end,
            set = function(_, value)
                ONSRaidTools.db.global.options.scale = value
                ONSRaidTools.mainWindow:SetScale(value)
            end
        },
    }
}


function ONSRaidTools:LoadOptionsValues()
    for _, option in pairs(ONSOptionsTable.args) do
        if option.get and option.set then
            local value = option.get()
            if value ~= nil then
                option.set(nil, value)
            end
        end
    end
end

function ONSRaidTools:OnEnable()
    ONSRaidTools:RegisterEvent("MODIFIER_STATE_CHANGED")
    ONSRaidTools:RegisterEvent("CHAT_MSG_ADDON")
    AceConfig:RegisterOptionsTable(AddOnName, ONSOptionsTable)
    self.optionsFrame = AceConfigDialog:AddToBlizOptions(AddOnName, AddOnName)
    C_ChatInfo.RegisterAddonMessagePrefix(AddOnName)
end

function ONSRaidTools:OnInitialize()
    self.db = LibStub("AceDB-3.0"):New("ONSRaidToolsDB", defaults)

    LDBIcon:Register(AddOnName, ONSRaidToolsBroker, self.db.global.options.minimap)
end

function ONSRaidTools:UpdateMinimapButton()
    if (self.db.global.options.minimap.hide) then
        LDBIcon:Hide(AddOnName)
    else
        LDBIcon:Show(AddOnName)
    end
end

function ONSRaidTools:ToggleMinimapButton()
    if (self.db.global.options.minimap.hide) then
        self.db.global.options.minimap.hide = false
    else
        self.db.global.options.minimap.hide = true
    end
    self:UpdateMinimapButton()
end

function ONSRaidTools:OpenSettings()
    if InterfaceOptionsFrame ~= nil then
        InterfaceOptionsFrame:Show()
    end
    InterfaceOptionsFrame_OpenToCategory(self.optionsFrame)
end

function ONSRaidTools:Print(msg)
    if not msg then
        return
    end
    local prefix = "|cff00ff00ONSRaidTools|r: "
    DEFAULT_CHAT_FRAME:AddMessage(prefix .. msg)
end

function ONSRaidTools:GetBossNameByModuleAndIndex(moduleName, index)
    if not moduleName or not index then return end
    index = tonumber(index)
    local module = self.modules[moduleName]
    if not module then return end
    local bosses = module.bosses
    if not bosses then return end
    if not bosses[index] then return end
    return bosses[index].name
end

function ONSRaidTools:CHAT_MSG_ADDON(event, prefix, msg)
    if prefix ~= AddOnName then return end
    local bossNum, imageNum = strsplit(",", msg)
    if not bossNum or not imageNum then return end
    bossNum = tonumber(bossNum)
    imageNum = tonumber(imageNum)
    local bossName = ONSRaidTools:GetBossNameByModuleAndIndex(ONSRaidTools.activeRaid, bossNum)
    if not bossName then
        bossName = "Boss" .. tostring(bossNum)
    end
    self:Print(string.format("Received image %d for %s", imageNum, bossName))

    self:LoadEncounter(bossNum, ONSRaidTools.activeRaid, imageNum)
end

-- /ons send <bossnum> <imagenum>
function ONSRaidTools:SendImageToRaid(bossNum, imageNum)
    local isLeader = UnitIsGroupLeader("player", LE_PARTY_CATEGORY_HOME)
    if not isLeader then
        self:Print("You need to be the raid leader to send images")
        return
    end
    if not bossNum then self:Print("No boss number provided") end
    if not imageNum then self:Print("No image number provided") end

    local bossName = ONSRaidTools:GetBossNameByModuleAndIndex(ONSRaidTools.activeRaid, bossNum)
    if not bossName then
        bossName = "Boss" .. tostring(bossNum)
    end
    self:Print(string.format("Sending image %d for %s", imageNum, bossName))
    local msg = string.format("%s,%s", bossNum, imageNum)
    C_ChatInfo.SendAddonMessage(AddOnName, msg, "RAID")
end

SLASH_ONS1 = "/ons"

SlashCmdList["ONS"] = function(msg)
    if msg == "settings" or msg == "config" then
        ONSRaidTools:OpenSettings()
    elseif string.match(msg, "send") then
        local _, _, bossnum, imagenum = string.find(msg, "send%s+(%d+)%s+(%d+)")
        if not bossnum or not imagenum then
            ONSRaidTools:Print("Usage: /ons send <bossnum> <imagenum>")
            return
        end
        ONSRaidTools:SendImageToRaid(bossnum, imagenum)
    elseif msg == "minimap" then
        ONSRaidTools:ToggleMinimapButton()
    elseif msg == "DFS1" or msg == "DFS2" then
        ONSRaidTools:Print(string.format("Loaded module %s", msg))
        ONSRaidTools.activeRaid = msg
        ONSRaidTools:SetActiveRaidToSelectView()
    else
        ONSRaidTools:ToggleFrame()
    end
end


images/df_s1.lua
--------------------------------------------
-- Dragonflight S1 - Vault of the Incarnates
--------------------------------------------
local AddOnName, _ = ...
local core = LibStub("AceAddon-3.0"):GetAddon(AddOnName)
local module = core:NewModule("DFS1")



module.bosses = {
    [1] = {
        ["icon"] = 4630361,
        ["name"] = "Eranog"
    },
    [2] = {
        ["icon"] = 4630366,
        ["name"] = "Terros"
    },
    [3] = {
        ["icon"] = 4630359,
        ["name"] = "Council"
    },
    [4] = {
        ["icon"] = 4630365,
        ["name"] = "Sennarth"
    },
    [5] = {
        ["icon"] = 4630367,
        ["name"] = "Dathea"
    },
    [6] = {
        ["icon"] = 4630362,
        ["name"] = "Kurog"
    },
    [7] = {
        ["icon"] = 4630360,
        ["name"] = "Diurna"
    },
    [8] = {
        ["icon"] = 4630364,
        ["name"] = "Raszageth"
    },
}



module.images = {
    [1] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\1_1_Boss.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\1_2_Boss.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\1_3_Boss.png",
    },
    [3] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\3_1_Boss.png",
    },
    [4] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\4_1_Boss.png",
    },
    [5] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\5_1_Boss.png",
    },
    [7] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\7_1_Boss.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s1\\7_2_Boss.png",
    }
}


images/df_s2.lua
--------------------------------------------------
-- Dragonflight S2 - Aberrus, The Shadow Crucible
--------------------------------------------------
local AddOnName, _ = ...
local core = LibStub("AceAddon-3.0"):GetAddon(AddOnName)
local module = core:NewModule("DFS2")

-- Icons are taken from the PTR might need reviewing when the patch is live.

module.bosses = {
    [1] = {
        ["icon"] = 5161745,
        ["name"] = "Kazzara"
    },
    [2] = {
        ["icon"] = 5161744,
        ["name"] = "Molgoth"
    },
    [3] = {
        ["icon"] = 5161744,
        ["name"] = "Dracthyr"
    },
    [4] = {
        ["icon"] = 5161751,
        ["name"] = "Zaqali"
    },
    [5] = {
        ["icon"] = 5161749,
        ["name"] = "Rashok"
    },
    [6] = {
        ["icon"] = 5161752,
        ["name"] = "Zskarn"
    },
    [7] = {
        ["icon"] = 5161746,
        ["name"] = "Magmorax"
    },
    [8] = {
        ["icon"] = 5161747,
        ["name"] = "Neltharion"
    },
    [9] = {
        ["icon"] = 5161750,
        ["name"] = "Sarkareth"
    },
}



-- Placeholder Images - File names will stay the same I will just update the artwork for each boss later.
module.images = {
    [1] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\1_1_Kazzara.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\1_2_Kazzara.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\1_3_Kazzara.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\1_4_Kazzara.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\1_5_Kazzara.png",
    },
    [2] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\2_1_Molgoth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\2_2_Molgoth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\2_3_Molgoth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\2_4_Molgoth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\2_5_Molgoth.png",
    },
    [3] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\3_1_Dracthyr.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\3_2_Dracthyr.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\3_3_Dracthyr.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\3_4_Dracthyr.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\3_5_Dracthyr.png",
    },
    [4] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\4_1_Zaqali.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\4_2_Zaqali.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\4_3_Zaqali.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\4_4_Zaqali.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\4_5_Zaqali.png",
    },
    [5] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\5_1_Rashok.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\5_2_Rashok.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\5_3_Rashok.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\5_4_Rashok.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\5_5_Rashok.png",
    },
    [6] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\6_1_Zskarn.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\6_2_Zskarn.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\6_3_Zskarn.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\6_4_Zskarn.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\6_5_Zskarn.png",
    },
    [7] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\7_1_Magmorax.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\7_2_Magmorax.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\7_3_Magmorax.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\7_4_Magmorax.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\7_5_Magmorax.png",
    },
    [8] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\8_1_Neltharion.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\8_2_Neltharion.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\8_3_Neltharion.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\8_4_Neltharion.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\8_5_Neltharion.png",
    },
    [9] = {
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\9_1_Sarkareth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\9_2_Sarkareth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\9_3_Sarkareth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\9_4_Sarkareth.png",
        "Interface\\Addons\\ONSRaidTools\\images\\df_s2\\9_5_Sarkareth.png",
    },
}


frames/components/Tabs.lua
local AddOnName, components = ...
-- tabsInfo = {
--     {
--         label = "Tab 1",
--         callback = function() print("Tab 1") end
--     },
--     {
--         label = "Tab 2",
--         callback = function() print("Tab 2") end
--           onLeave = function() print("Tab 2 onLeave") end
--     }
-- }
local tabsFramePool = CreateFramePool("Button", nil, "ONSRaidToolsTabTemplate")

function components:SetActiveTab(frames, index)
    -- frames -> all the tab frames
    -- index -> the index of the tab to set active
    for i, j in ipairs(frames) do
        if i == index then
            j:SetActive()
        else
            j:SetInactive()
        end
    end
end

function components:CreateTabs(tabsInfo)
    local frames = {}
    self:KillTabs() -- We only want one set of tabs at a time
    for i, tabInfo in ipairs(tabsInfo) do
        local tab = tabsFramePool:Acquire()
        tab.text:SetText(i)
        tab:SetScript("OnClick", function()
            self:SetActiveTab(frames, i)
            tabInfo.callback()
        end)

        -- tab:SetScript("OnEnter", function(f)
        --     GameTooltip:SetOwner(f, "ANCHOR_BOTTOM", 0, -1)
        --     GameTooltip:SetText(tabInfo.label)
        --     GameTooltip:Show()
        -- end)
        tab:SetScript("OnLeave", function(f)
            -- GameTooltip:Hide()
            tabInfo.onLeave()
        end)
        table.insert(frames, tab)
    end
    return frames
end

function components:LayoutTabs(tabs, parentFrame, gap, paddingX, paddingBottom)
    local tabWidth = (parentFrame:GetWidth() - (paddingX * 2) - ((#tabs - 1) * gap)) / #tabs
    for i, tab in ipairs(tabs) do
        tab:SetParent(parentFrame)
        tab:ClearAllPoints()
        tab:SetPoint("LEFT", parentFrame, "LEFT", paddingX + ((i - 1) * (tabWidth + gap)), paddingBottom)
        tab:SetWidth(tabWidth)
        tab:Show()
    end
    parentFrame.tabs = tabs
end

function components:KillTabs()
    tabsFramePool:ReleaseAll()
end


mainWindow.lua
local AddOnName, components = ...
local ONSRaidTools = LibStub("AceAddon-3.0"):GetAddon(AddOnName)

-- TODO: Add recipe module

local VIEWS = {
    IMAGE = "image",
    SELECT = "select"
}


local moveCrosshair = "Interface/CURSOR/UI-Cursor-Move.crosshair"
local function createWindow()
    -- Create the frame
    local frame = CreateFrame("Frame", AddOnName .. "mainWindow", UIParent, "ONSImagViewMainWindowTemplate")
    return frame
end


function ONSRaidTools:HideImageViewButtons()
    if not self.imageView then return end
    self.imageView.tabsHolder:Hide()
    self.imageView.menuButton:Hide()
end

function ONSRaidTools:ShowImageViewButtons()
    if not self.imageView then return end
    self.imageView.tabsHolder:Show()
    self.imageView.menuButton:Show()
end

local function CheckFrameMouseover(f)
    if not f then return end
    if f:IsMouseOver() then
        return true
    end
    local children = { f:GetChildren() }
    while #children > 0 do
        local child = table.remove(children, 1)
        if child:IsMouseOver() then
            return true
        end
        local subChildren = { child:GetChildren() }
        for i = 1, #subChildren do
            table.insert(children, subChildren[i])
        end
    end
    return false
end


function ONSRaidTools:UpdateImageViewMouseOverState()
    if not self.imageView then return end
    if CheckFrameMouseover(self.imageView) then
        self:ShowImageViewButtons()
    else
        self:HideImageViewButtons()
    end
end

function ONSRaidTools:AddListernersToView(view)
    if not view then return error("view is nil") end
    view:SetScript("OnEnter", function(f)
        if IsShiftKeyDown() then
            f:EnableMouse(false)
            f:SetAlpha(0.1)
            return
        end
        if view.name == VIEWS.IMAGE then
            self:UpdateImageViewMouseOverState()
        end
    end)
    view:SetScript("OnLeave", function(f)
        local v = self:GetCurrentView()
        if v then
            v:SetAlpha(1)
        end
        if view.name == VIEWS.IMAGE then
            self:UpdateImageViewMouseOverState()
        end
    end)
end

function ONSRaidTools:GetCurrentView()
    if not self.activeView then return end
    local current = self:GetViewByName(self.activeView)
    return current
end

function ONSRaidTools:AddMainListeners()
    local frame = self.mainWindow
    frame:SetScript("OnDragStart", function(f)
        f:StartMoving()
        f:SetUserPlaced(false)
        SetCursor(moveCrosshair)
    end)

    frame:SetScript("OnDragStop", function(f)
        f:StopMovingOrSizing()
        SetCursor(nil)
        -- Get the position of the frame
        local point, _, relativePoint, xOfs, yOfs = f:GetPoint()
        -- If we found a point then save that position
        if point and relativePoint and xOfs and xOfs then
            self.db.global.position = {
                point = point,
                relativePoint = relativePoint,
                x = xOfs,
                y = yOfs
            }
        end
    end)

    frame:SetScript("OnLeave", function(f)
        local view = self:GetCurrentView()
        if view then
            -- view:EnableMouse(true)
            view:SetAlpha(1)
        end
    end)
end

function ONSRaidTools:SetInitPosition()
    local x, y, point, relativePoint
    if self.db.global.position then
        x = self.db.global.position.x
        y = self.db.global.position.y
        relativePoint = self.db.global.position.relativePoint
        point = self.db.global.position.point
    end
    local frame = self.mainWindow
    frame:SetPoint(point, UIParent, relativePoint, x, y)
end

local function createImageView()
    local frame = CreateFrame("Frame", AddOnName .. "ImageView", UIParent, "ONSRaidToolsTemplate")
    return frame
end

function ONSRaidTools:GetViewByName(name)
    if name == VIEWS.IMAGE then
        if not self.imageView then
            self:InitImageView()
        end
        return self.imageView
    elseif name == VIEWS.SELECT then
        if not self.selectView then
            self:InitSelectView()
        end
        return self.selectView
    end
end

function ONSRaidTools:SetView(view)
    if not view then return end
    if not self.activeView then return end
    local current = self:GetCurrentView()
    if current then
        current:Hide()
    end
    view:SetParent(self.mainWindow)
    view:SetPoint("TOPLEFT", self.mainWindow, "TOPLEFT", 0, 0)
    view:SetPoint("BOTTOMRIGHT", self.mainWindow, "BOTTOMRIGHT", 0, 0)
    view:Show()
    self.activeView = view.name
end

function ONSRaidTools:LoadImageToView(img)
    -- img -> path to image
    if not img then return end
    self.imageView.img:SetTexture(img)
end

function ONSRaidTools:CreateImageViewTabs(tabsTable, overrideIntitial)
    if not tabsTable then return end
    local tabs = components:CreateTabs(tabsTable)
    components:LayoutTabs(tabs, self.imageView.tabsHolder, 5, 5, 4)
    local indexToLoad = 1
    if overrideIntitial then
        indexToLoad = overrideIntitial
    end
    -- Set the first tab as active
    tabsTable[indexToLoad].callback()
    components:SetActiveTab(tabs, indexToLoad)
end

function ONSRaidTools:LoadCurrentImageCollection(overrideIntitial)
    if not self.db.global.loadedEncounter then return end
    if not self.db.global.loadedEncounter.images then return end

    local images = self.db.global.loadedEncounter.images
    local tabsTable = {}

    for i, path in ipairs(images) do
        local tab = {
            label = "Image " .. i,
            callback = function()
                self:LoadImageToView(path)
            end,
            onLeave = function()
                self:UpdateImageViewMouseOverState()
            end
        }
        table.insert(tabsTable, tab)
    end
    self:CreateImageViewTabs(tabsTable, overrideIntitial)
end

function ONSRaidTools:SetEncounterNameByIndex(index)
    if not index then return end
    local bossName = self:GetBossNameByModuleAndIndex(self.activeRaid, index)
    if not bossName then return end
    self.imageView.menuButton.bossName:SetText(bossName)
end

-- Example: ONSRaidTools:LoadEncounter(1, "dfs1")
function ONSRaidTools:LoadEncounter(encounterIndex, moduleName, overrideIntitial)
    -- encounterIndex -> index of encounter in the raid
    if not encounterIndex then
        error("encounterIndex not found")
        return
    end
    if not moduleName then
        error("moduleName not found")
        return
    end
    local module = self.modules[moduleName]
    if not module then
        error("module not found")
        return
    end
    local images = module.images[encounterIndex]
    if not images then
        error("images not found")
        return
    end
    local info = module.bosses[encounterIndex]
    if not info then
        error("info not found")
        return
    end
    self.db.global.loadedEncounter.images = images
    self.db.global.loadedEncounter.info = info
    self:SetEncounterNameByIndex(encounterIndex)
    self:LoadCurrentImageCollection(overrideIntitial)
end

function ONSRaidTools:InitImageView()
    self.imageView = createImageView()
    self.imageView.menuButton:SetScript("OnClick", function(f)
        if not self.selectView then
            self:InitSelectView()
        end
        self:SetView(self.selectView)
    end)


    self.imageView.menuButton:SetScript("OnLeave", function(f)
        self:UpdateImageViewMouseOverState()
    end)

    self.imageView.tabsHolder:SetScript("OnLeave", function(f)
        self:UpdateImageViewMouseOverState()
    end)


    self.imageView.name = VIEWS.IMAGE
    self:AddListernersToView(self.imageView)
end

local function createSelectView()
    local frame = CreateFrame("Frame", nil, UIParent, "ONSSelectViewTemplate")
    return frame
end


function ONSRaidTools:SetActiveRaidToSelectView()
    -- only for hardcoded raid
    -- Set the module to the module with the specified name
    local module = self.modules[self.activeRaid]
    -- Check if the module is set, if not return
    if not module then
        error("module not found")
        return
    end
    for i, button in ipairs(self.selectView.encounterButtons) do
        button:Hide()
    end
    local buttonIndex = 1
    for bossIndex, bossInfo in ipairs(module.bosses) do
        if module.images[bossIndex] then
            local button = self.selectView.encounterButtons[buttonIndex]
            button.icon:SetTexture(bossInfo.icon)
            button.label:SetText(bossInfo.name)
            button:SetScript("OnClick", function(f)
                self:LoadEncounter(bossIndex, self.activeRaid)
                self:SetView(self.imageView)
            end)
            button:Show()
            buttonIndex = buttonIndex + 1
        end
    end
end

function ONSRaidTools:InitSelectView()
    self.selectView = createSelectView()
    self:AddListernersToView(self.selectView)
    self.selectView.titleBar.backButton:SetScript("OnClick", function(f)
        self:SetView(self.imageView)
    end)

    self.selectView.titleBar.settingsButton:SetScript("OnClick", function(f)
        self:OpenSettings()
    end)

    local gap = 5
    local row = 0
    local max = 14
    self.selectView.encounterButtons = {}

    for buttonIndex = 1, max do
        local button = CreateFrame("Button", nil, self.selectView.encounterButtonHolder,
            "ONSRaidToolsEncounterButtonTemplate")
        local width = button:GetWidth()
        local height = button:GetHeight()
        local col = (buttonIndex % 2 == 0 and 2 or 1)
        if col == 1 then
            row = row + 1
        end

        local xOffset = (col - 1) * (width + gap)
        local yOffset = -((row - 1) * (height + gap))

        button:SetPoint("TOPLEFT", self.selectView.encounterButtonHolder, "TOPLEFT", xOffset, yOffset)
        button:Hide()
        table.insert(self.selectView.encounterButtons, button)
    end

    self:SetActiveRaidToSelectView()
    self.selectView.name = VIEWS.SELECT
end

function ONSRaidTools:setupMainWindow()
    self.mainWindow = createWindow()
    self:AddMainListeners()
    self:SetInitPosition()
    ONSRaidTools:GetViewByName(VIEWS.IMAGE)
    self.activeView = VIEWS.IMAGE
    self:SetView(self.imageView)
    self:LoadOptionsValues()
end

function ONSRaidTools:ToggleFrame()
    --Create the window if it doesn't exist
    if not self.mainWindow then
        self:setupMainWindow()
    end

    --Show or hide the window depending on its current state
    if self.mainWindow:IsShown() then
        self.mainWindow:Hide()
    else
        self.mainWindow:Show()
    end
end

function ONSRaidTools:MODIFIER_STATE_CHANGED(e, key, state)
    if key ~= "LSHIFT" and key ~= "RSHIFT" then return end
    if not self.imageView then return end
    if state == 1 and (self.imageView:IsMouseOver(1, 0, 1, 0)) then
        self.mainWindow.moveInfo:Show()
        self.imageView:EnableMouse(false)
        self.imageView:SetAlpha(0.1)
    else
        self.mainWindow.moveInfo:Hide()
        self.imageView:EnableMouse(true)
        self.imageView:SetAlpha(1)
        if (self.imageView:IsMouseOver(1, 0, 1, 0)) then
            self.imageView.tabsHolder:Show()
            self.imageView.menuButton:Show()
        end
    end

    if not self.selectView then return end
    if state == 1 and (self.selectView:IsMouseOver(1, 0, 1, 0)) then
        self.mainWindow.moveInfo:Show()
        self.selectView:EnableMouse(false)
        self.selectView:SetAlpha(0.1)
    else
        self.mainWindow.moveInfo:Hide()
        self.selectView:EnableMouse(true)
        self.selectView:SetAlpha(1)
    end
end
